### PR TITLE
feat: improve player trace

### PR DESCRIPTION
**Pull request by RainbowwPhoenixx**
- greatly reduced ram usage
- support for multiple traces
- more control over drawing of the traces
- more accurate speed deltas

### CODE REVIEW

File: src/Features/PlayerTrace.cpp [hunk 1]
```diff
@@ -20,9 +20,13 @@ PlayerTrace* playerTrace;
...
 #endif
 
-Variable sar_player_trace_draw("sar_player_trace_draw", "0", "Display the recorded player trace\n");
 Variable sar_player_trace_record("sar_player_trace_record", "0", "Record player trace while set to 1\n");
 Variable sar_player_trace_size("sar_player_trace_size", "0", 0, "Maximum size of the player trace (in ticks), 0 for unlimited\n");
+Variable sar_player_trace_autoclear("sar_player_trace_autoclear", "1", "Automatically clear the trace on session start\n");
+
+Variable sar_player_trace_draw("sar_player_trace_draw", "0", "Display the recorded player trace. Requires cheats\n");
+Variable sar_player_trace_draw_through_walls("sar_player_trace_draw_through_walls", "1", "Display the player trace through walls. Requires sar_player_trace_draw");
+Variable sar_player_trace_draw_speed_deltas("sar_player_trace_draw_speed_deltas", "1", "Display the speed deltas. Requires sar_player_trace_draw");
```
Comment (mlugg) [inline src/Features/PlayerTrace.cpp:27]: Missing newlines on cvar descriptions

File: src/Features/PlayerTrace.cpp [hunk 2]
```diff
@@ -147,12 +152,14 @@ ON_EVENT(PRE_TICK) {
...
 ON_EVENT(SESSION_START) {
-    playerTrace->Clear();
+    if (sar_player_trace_autoclear.GetBool())
+        playerTrace->Clear();
 }
 
 HUD_ELEMENT2(player_trace_draw_speed, "1", "Toggles drawing the speed deltas on the player traces\n", HudType_InGame) {
 
     if (!sar_player_trace_draw.GetBool()) return;
+    if (!sar_player_trace_draw_speed_deltas.GetBool()) return;
```
Comment (mlugg) [inline src/Features/PlayerTrace.cpp:187]: If you have your own toggle for this, you should use HUD_ELEMENT2_NO_DISABLE, as else you'll have a `sar_hud_player_trace_draw_speed` toggle too (which ofc makes no sense lol)

File: src/Features/PlayerTrace.cpp [hunk 3]
```diff
@@ -79,48 +96,40 @@ void PlayerTrace::DrawInWorld(float time) const
...
+    Vector screen_pos;
+
+    if (begin_tick == 0) begin_tick += idx_offset;
+    Vector speed_a = (trace[begin_tick - idx_offset] - trace[begin_tick])/idx_offset;
+    Vector speed_b = (trace[end_tick - idx_offset] - trace[end_tick])/idx_offset;
+    float speed_delta = speed_b.Length2D() - speed_a.Length2D();
+
+    engine->PointToScreen(trace[begin_tick + (end_tick - begin_tick)/2] + hud_offset, screen_pos);
+    ctx->DrawElementOnScreen((*hud_id)++, screen_pos.x, screen_pos.y, "%10.2f", speed_delta * 60);
+}
```
Comment (mlugg) [inline src/Features/PlayerTrace.cpp:151]: there are a lot of words to read in this function and i am very sleepy so i am going to assume all your maths is perfect

File: src/Features/PlayerTrace.hpp [hunk 4]
```diff
@@ -1,27 +1,56 @@
...
 
+class PlayerTrace : public Feature {
+private:
+    // In order to arbitrarily number traces
+    std::map<size_t, Trace> traces;
 public:
     PlayerTrace();
     // Add a point to the player trace
-    void AddPoint(const Vector Point);
+    void AddPoint(const size_t trace_idx, void *player);
```
Comment (mlugg) [inline src/Features/PlayerTrace.hpp:52]: a plain (non-pointer, non-reference) argument doesn't really need to be const-qualified, i honestly didn't even know you COULD do that

File: src/Features/PlayerTrace.cpp (hunk 1 above)
Comment (RainbowwPhoenixx) [reply src/Features/PlayerTrace.cpp:27]: Considering I rewrote basically every help string, that's very funny

**Pull request merged**